### PR TITLE
Cuckoo consensus engine

### PR DESCRIPTION
**Pull request by Kais-DkM**

### CODE REVIEW

File: core/src/consensus/cuckoo/mod.rs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+impl ConsensusEngine<CodeChainMachine> for Cuckoo {
+    fn name(&self) -> &str {
+        "Cuckoo"
+    }
+
+    fn machine(&self) -> &CodeChainMachine {
+        &self.machine
+    }
+
+    /// The number of additional header fields required for this engine.
```
Comment (kseo) [inline core/src/consensus/cuckoo/mod.rs]: Let's remove the comment.

File: core/src/consensus/cuckoo/mod.rs [hunk 2]
```diff
@@ -17,14 +17,41 @@
...
 use self::params::CuckooParams;
 use super::super::block::ExecutedBlock;
 use super::super::codechain_machine::CodeChainMachine;
-use super::super::error::Error;
+use super::super::error::{BlockError, Error};
 use super::super::header::Header;
 use super::ConsensusEngine;
 
+
+/// Ethash specific seal
```
Comment (kseo) [inline core/src/consensus/cuckoo/mod.rs]: Ethash -> Cuckoo

File: core/src/consensus/cuckoo/mod.rs [hunk 3]
```diff
@@ -17,14 +17,41 @@
...
+#[derive(Debug, PartialEq)]
+pub struct Seal {
+    pub nonce: u64,
+    pub proof: Vec<u32>,
+}
+
+impl Seal {
+    /// Tries to parse rlp as ethash seal.
```
Comment (kseo) [inline core/src/consensus/cuckoo/mod.rs]: ethash -> cuckoo

File: core/src/consensus/cuckoo/mod.rs [hunk 4]
```diff
@@ -54,12 +55,16 @@ impl Seal {
 
 pub struct Cuckoo {
     machine: CodeChainMachine,
+    verifier: CuckooVerifier,
 }
 
 impl Cuckoo {
     pub fn new(params: CuckooParams, machine: CodeChainMachine) -> Self {
+        // get parameters from params
```
Comment (kseo) [inline core/src/consensus/cuckoo/mod.rs]: Add FIXME: note.

File: core/src/consensus/cuckoo/mod.rs [hunk 5]
```diff
@@ -82,19 +87,37 @@ impl ConsensusEngine<CodeChainMachine> for Cuckoo {
     }
 
     fn verify_block_basic(&self, header: &Header) -> Result<(), Error> {
-        unimplemented!()
+        let seal = Seal::parse_seal(header.seal())?;
+        // FIXME: check minimum difficulty
```
Comment (kseo) [inline core/src/consensus/cuckoo/mod.rs]: Leave `unimplemented!()` here.

**Comment by sgkim126** [comment]
It seems that it's because 1943b214ae0ff4968ff931654e9ecde857a5e580 upgrades the clippy version from 0.0.198 to 0.0.211.

**Comment by sgkim126** [comment]
FYI, `cargo build` succeeds but `cargo fetch` fails.

**Comment by sgkim126** [comment]
It was because secp256k1 has an optional dependency on clippy.
CodeChain uses the secp256k1 version 0.5.7, but the latest(0.9.2) doesn't depend on that.
So I think upgrading secp256k1 solves the problem.

**Comment by sgkim126** [comment]
@kseo By the way, is there a reason that CodeChain holds the secp256k1 source? Why don't we add a dependency to that crate?

File: core/res/cuckoo.json [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+          "blockReward": "0x0d",
+          "minScore" : "0x020000",
+          "maxVertex" : "0x40000000",
+          "maxEdge" : "0x20000000",
+          "cycleLength" : "42"
+      }
+    }
+  },
+  "params": {
+    "maxExtraDataSize": "0x20",
```
Comment (kseo) [inline core/res/cuckoo.json]: Add maxMetadataSize.

File: json/src/spec/cuckoo.rs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    use ctypes::U256;
+    use serde_json;
+
+    use super::super::super::uint::Uint;
+    use super::*;
+
+    #[test]
+    fn cuckoo_deserialization() {
+        let s = r#"{
+			"params": {
```
Comment (kseo) [inline json/src/spec/cuckoo.rs]: Fix indentation.

File: json/src/spec/engine.rs [hunk 8]
```diff
@@ -91,5 +93,22 @@ mod tests {
             Engine::Tendermint(_) => {} // Tendermint is unit tested in its own file.
             _ => panic!(),
         };
+
+        let s = r#"{
+			"cuckoo": {
```
Comment (kseo) [inline json/src/spec/engine.rs]: Fix indentation.

Comment (Kais-DkM) [reply json/src/spec/engine.rs]: Other test codes in same file has similar problem. I'll patch this in master branch manually.

**Pull request merged**